Poor error message with partial application of function arrow

If I write this:

```hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE GADTs #-}
{-# LANGUAGE PolyKinds #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE StandaloneKindSignatures #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE TypeApplications #-}
{-# LANGUAGE TypeFamilies #-}
module Bug where

import Data.Proxy
import Data.Proxy.Singletons
import Data.Singletons.TH

$(singletons [d|
  data Foo = MkFoo (Proxy (->))
  |])
```

Then `singletons-th` will produce this error:

```
[1 of 1] Compiling Bug              ( Dyn.hs, interpreted )

Dyn.hs:15:2: error: Q monad failure
   |
15 | $(singletons [d|
   |  ^^^^^^^^^^^^^^^...

Dyn.hs:15:2: error:
    Illegal Haskell construct encountered:
        headed by: DArrowT
        applied to: []

   |
15 | $(singletons [d|
   |  ^^^^^^^^^^^^^^^...
```

This is a poor error message, as there's certainly nothing illegal about the program we're trying to promote. It may be the case that `singletons-th` can't successfully handle something like this, but if so, the error message should just state that.

This was originally spotted in #510, but this issue is separate from the main problem being discussed in #510.

--- a/README.md
+++ b/README.md
@@ -1277,6 +1277,7 @@ The following constructs are either unsupported or almost never work:
 * `TypeApplications`
 * Irrefutable patterns
 * `{-# UNPACK #-}` pragmas
+* partial application of the `(->)` type
 
 See the following sections for more details.
 
@@ -1481,3 +1482,10 @@ singled data types represent their argument types using existential type
 variables, and any data constructor that explicitly uses existential
 quantification cannot be unpacked. See
 [GHC#10016](https://gitlab.haskell.org/ghc/ghc/-/issues/10016).
+
+### Partial application of the `(->)` type
+
+`singletons-th` can only promote `(->)` when it is applied to exactly two
+arguments. Attempting to promote `(->)` to zero or one argument will result in
+an error. As a consequence, it is impossible to promote instances like the
+`Functor ((->) r)` instance, so `singletons-base` does not provide them.

--- a/singletons-base/tests/SingletonsBaseTestSuite.hs
+++ b/singletons-base/tests/SingletonsBaseTestSuite.hs
@@ -143,6 +143,7 @@ tests =
     , compileAndDumpStdTest "T489"
     , compileAndDumpStdTest "T492"
     , compileAndDumpStdTest "Natural"
+    , compileAndDumpStdTest "T511"
     ],
     testCompileAndDumpGroup "Promote"
     [ compileAndDumpStdTest "Constructors"

--- a/singletons-base/tests/compile-and-dump/Singletons/T511.golden
+++ b/singletons-base/tests/compile-and-dump/Singletons/T511.golden
@@ -1,0 +1,13 @@
+
+Singletons/T511.hs:0:0: error: Q monad failure
+  |
+7 | $(singletons [d|
+  |  ^^^^^^^^^^^^^^^...
+
+Singletons/T511.hs:0:0: error:
+    `singletons-th` does not support partial applications of (->)
+	In the type: Data.Proxy.Proxy (->)
+
+  |
+7 | $(singletons [d|
+  |  ^^^^^^^^^^^^^^^...

--- a/singletons-base/tests/compile-and-dump/Singletons/T511.hs
+++ b/singletons-base/tests/compile-and-dump/Singletons/T511.hs
@@ -1,0 +1,9 @@
+module T511 where
+
+import Data.Proxy
+import Data.Proxy.Singletons
+import Data.Singletons.TH
+
+$(singletons [d|
+  data Foo = MkFoo (Proxy (->))
+  |])

--- a/singletons-th/src/Data/Singletons/TH/Promote/Type.hs
+++ b/singletons-th/src/Data/Singletons/TH/Promote/Type.hs
@@ -94,9 +94,12 @@ promoteType_options pto typ = do
     go args     (DConT name) = do
       opts <- getOptions
       return $ applyDType (DConT (promotedDataTypeOrConName opts name)) args
-    go [DTANormal k1, DTANormal k2] DArrowT
-      = return $ DConT tyFunArrowName `DAppT` k1 `DAppT` k2
-    go args     ty@DArrowT = illegal args ty
+    go args     ty@DArrowT =
+      case filterDTANormals args of
+        []        -> noPartialArrows
+        [_]       -> noPartialArrows
+        [k1, k2]  -> return $ DConT tyFunArrowName `DAppT` k1 `DAppT` k2
+        (_:_:_:_) -> illegal args ty
     go []       ty@DLitT{} = pure ty
     go args     ty@DLitT{} = illegal args ty
     go args     ty@DWildCardT{}
@@ -106,8 +109,17 @@ promoteType_options pto typ = do
       = fail $ unlines
           [ "`singletons-th` does not support wildcard types"
           , "\tunless they appear in visible type patterns of data constructors"
-          , "\tIn the type: " ++ pprint (sweeten typ)
+          , "\t" ++ herald
           ]
+
+    noPartialArrows :: m a
+    noPartialArrows = fail $ unlines
+      [ "`singletons-th` does not support partial applications of (->)"
+      , "\t" ++ herald
+      ]
+
+    herald :: String
+    herald = "In the type: " ++ pprint (sweeten typ)
 
     illegal :: [DTypeArg] -> DType -> m a
     illegal args hd = fail $ unlines

